### PR TITLE
spider: preserve query parameters with same name when canonicalising URL

### PR DESCRIPTION
**Pull request by thc202**
Change class URLCanonicalizer to keep query parameters with same name
to preserve (or attempt to) the meaning of the original URL.
For example, with old behaviour URLs like:
 http://example.com/?array[]=1&array[]=2
 http://example.com/?IDs=1&IDs=2
would loose one of the query parameters, which could result in different
content being served.
Add tests to assert the expected behaviour.

### CODE REVIEW

**Comment by kingthorin** [comment]
Wasn't there a thread in the user group or an issue related to this?

File: src/org/zaproxy/zap/spider/URLCanonicalizer.java [hunk 1]
```diff
@@ -446,4 +450,83 @@ private static String normalizePath(final String path) {
...
+			}
+			int result = name.compareTo(other.name);
+			if (result != 0) {
+				return result;
+			}
+			return value.compareTo(other.value);
+		}
+
+		@Override
+		public int hashCode() {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/spider/URLCanonicalizer.java]: I think I saw this code in another PR this week, should it be standardized in a utility class?

**Comment by thc202** [comment]
Related to spider? I'm not recalling any.
(There are issues related to mishandling of parameters with same name in the active scanner, maybe were those?)

**Comment by kingthorin** [comment]
Could be... no big deal.
I'm good with this change, just the one question which isn't really a blocking issue.

File: src/org/zaproxy/zap/spider/URLCanonicalizer.java (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/spider/URLCanonicalizer.java]: The hashCode() implementation?
(NameValuePair perhaps?)

**Comment by thc202** [comment]
Added a comment to your question.

Updated the class QueryParameter to remove redundant null checks.

File: src/org/zaproxy/zap/spider/URLCanonicalizer.java (hunk 1 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/spider/URLCanonicalizer.java]: hashCode()....it was over here: https://github.com/thc202/zaproxy/commit/f2fb322ad21f0a5072accb0663426df18c2106c3#diff-24025444157a86c23540fcbb1c660144R106

Comment (thc202) [inline src/org/zaproxy/zap/spider/URLCanonicalizer.java]: Right, there are not much to standardise unfortunately, it really depends on the state of each class.
There are utility classes that help reduce "boiler plate" code though (e.g. Commons Lang's HashCodeBuilder).

Comment (kingthorin) [inline src/org/zaproxy/zap/spider/URLCanonicalizer.java]: Fair enough.

**Comment by kingthorin** [comment]
Looks good

**Pull request merged**